File: reservation-frontend/src/app/core/services/espace.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment.development';

import { Espace } from '../models/espace/espace';

@Injectable({
  providedIn: 'root'
})
export class EspaceService {
  
  // URL de ton API Spring Boot
  private readonly apiUrl = `${environment.apiUrl}/api/espaces`;

  constructor(private http: HttpClient) {}

  // POST /api/espaces
  creerEspace(espace: Espace): Observable<Espace> {
    return this.http.post<Espace>(this.apiUrl, espace);
  }

  // GET /api/espaces
  listerTousLesEspaces(): Observable<Espace[]> {
    return this.http.get<Espace[]>(this.apiUrl);
  }

  // GET /api/espaces/{id}
  trouverEspaceParId(id: number): Observable<Espace> {
    return this.http.get<Espace>(`${this.apiUrl}/${id}`);
  }

  // PUT /api/espaces/{id}
  modifierEspace(id: number, espaceDetails: Espace): Observable<Espace> {
    return this.http.put<Espace>(`${this.apiUrl}/${id}`, espaceDetails);
  }

  // DELETE /api/espaces/{id}
  supprimerEspace(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  // GET /api/espaces/ville/{ville}
  trouverEspacesParVille(ville: string): Observable<Espace[]> {
    return this.http.get<Espace[]>(`${this.apiUrl}/ville/${ville}`);
  }

  // GET /api/espaces/search?keyword={keyword}
  rechercherEspaces(keyword: string): Observable<Espace[]> {
    return this.http.get<Espace[]>(`${this.apiUrl}/search`, {
      params: { keyword }
    });
  }

  // ✅ AJOUT : GET /api/espaces/user/{userId} (Récupérer l'espace et ses salles par l'ID utilisateur)
  trouverEspaceParUserId(userId: number): Observable<Espace> {
    return this.http.get<Espace>(`${this.apiUrl}/user/${userId}`);
  }
} 

File: Reservation/src/main/java/net/travel/reservation/services/EspaceService.java
package net.travel.reservation.services;

import net.travel.reservation.entites.Espace;
import net.travel.reservation.entites.User;
import net.travel.reservation.repositories.EspaceRepository;
import net.travel.reservation.security.SecurityUtils; // Import de ton utilitaire de sécurité
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class EspaceService {

    @Autowired
    private EspaceRepository espaceRepository;

    @Autowired
    private SecurityUtils securityUtils; // Injection de SecurityUtils

    public Espace creerEspace(Espace espace) {
        if (espaceRepository.existsByNom(espace.getNom())) {
            throw new RuntimeException("Un espace avec ce nom existe déjà.");
        }
        return espaceRepository.save(espace);
    }

    public Espace modifierEspace(Long id, Espace espaceDetails) {
        Espace espace = espaceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Espace introuvable avec l'ID : " + id));

        espace.setNom(espaceDetails.getNom());
        espace.setDescription(espaceDetails.getDescription());
        espace.setAdresse(espaceDetails.getAdresse());
        espace.setVille(espaceDetails.getVille());
        espace.setTelephone(espaceDetails.getTelephone());

        return espaceRepository.save(espace);
    }

    public Optional<Espace> trouverEspaceParId(Long id) {
        return espaceRepository.findById(id);
    }

    public List<Espace> listerTousLesEspaces() {
        return espaceRepository.findAll();
    }

    public void supprimerEspace(Long id) {
        if (!espaceRepository.existsById(id)) {
            throw new RuntimeException("Impossible de supprimer : Espace introuvable.");
        }
        espaceRepository.deleteById(id);
    }

    // Recherche par ville (insensible à la casse)
    public List<Espace> trouverEspacesParVille(String ville) {
        return espaceRepository.findByVilleIgnoreCase(ville);
    }

    // Recherche par mot-clé (nom ou ville) pour les barres de recherche frontend
    public List<Espace> rechercherEspaces(String keyword) {
        return espaceRepository.searchByNomOrVille(keyword);
    }

    // ✅ MODIFICATION : Récupérer l'espace et ses salles de l'utilisateur connecté automatiquement
    public Optional<Espace> trouverEspaceParUtilisateurConnecte() {
        User currentUser = securityUtils.getCurrentUser();
        return espaceRepository.findByUserId(currentUser.getUserId());
    }
}